### ISSUE TITLE
Publishing?

### ISSUE THREAD

**Issue by Hoverbear** [pingcap/kvproto#280]
I'm interested in publishing this crate so that it can be used as part of the upcoming TiKV client in Rust (which will be published on crates.io).

This already builds cleanly on nightly and has no illegal dependencies, so there should be nothing blocking publishing this.

**Comment by Hoverbear** [comment]
@siddontang PTAL

**Comment by Hoverbear** [comment]
We need to seriously consider this with the `tikv-client` for Rust. We cannot publish the client without all of its dependencies also being published.

**Comment by nrc** [comment]
Some things we need to do/think about:

- [ ] Use a crates.io version of Raft (I think this just needs us to finish the Prost migration and other Raft versioning issues.)
- [ ] Consider the name - 'kvproto' is not very self-explanatory (tikv_proto is suggested above, I prefer tikv-proto (consistent with rust-client, which should probably be tikv-client.rs or something)
- [ ] Consider and document the process for releases
- [ ] What should a Rust crate look like? We probably should not include the Go and CPP code in the distributed tarball, for example.
- [ ] building - are we happy with the current process and requirements for building (see also #354)
- [ ] think about the project org maybe? (Some things worth considering: why is Raft the source of truth for it's Protos, but this project is for the other protos? Why would a proto be here vs elsewhere? Should we be committing the generated code? Is having a multi-language project the right approach (especially if we add more languages in the future)?

Probably more...

**Comment by Hoverbear** [comment]
@nrc Do you think extracting only the tikv related protos and making them a dependency here would be a good first step to enable this? Maybe a `tikv-api` crate? Then we could publish that.

**Comment by nrc** [comment]
> Do you think extracting only the tikv related protos and making them a dependency here would be a good first step to enable this? Maybe a tikv-api crate? 

I think that is probably a good idea. I wonder whether the Raft protos should be part of that?

**Comment by ice1000** [comment]
These protos were too big for Raft (which is relatively small) to depend on, but the situation may change after prost

**Comment by BusyJay** [comment]
No, raft is a standard library that should keep its own protocols.

**Comment by nrc** [comment]
I don't understand either of those points, sorry.

> These protos were too big for Raft (which is relatively small) to depend on

Why is size an issue?

> No, raft is a standard library that should keep its own protocols.

What do you mean by "standard" (and by implication why are TikV, etc., not standard?)

**Comment by BusyJay** [comment]
Sorry, it should be standalone. :rofl:

**Comment by ice1000** [comment]
> Why is size an issue?

Hmm, compiling kvproto takes a long time.

**Comment by nrc** [comment]
> Hmm, compiling kvproto takes a long time.

This is true of a full build (1m 16s for me), but building kvproto itself only takes 7s. Given that any deps of kvproto should be required by Raft too, it is that which matters (in practice, I think we pull in grpc-rs, which we don't really need to, I'll file an issue about that).

**Comment by nrc** [comment]
> Sorry, it should be standalone. 🤣

That makes more sense, thanks! However, I'm still a bit unclear *why* Raft is a standalone library in this sense. If it is meant to be self-contained, why do we need to re-export the proto interface here?

**Comment by BusyJay** [comment]
> ...why do we need to re-export the proto interface here?

Because kvproto uses the struct of raft protocol directly, and due to the limited of protobuf compiler, we need to reexport raft proto interface to make the generated code compiled.

**Comment by nrc** [comment]
> Because kvproto uses the struct of raft protocol directly, and due to the limited of protobuf compiler, we need to reexport raft proto interface to make the generated code compiled.

Cool, thanks!

**Comment by nrc** [comment]
Proposal:

* Move the Raft protos into a RaftProto repo
* Raft depends on RaftProto
* KvProto depends on RaftProto, not Raft
* Move the code for generating protos in Go into protobuf-build so that KvProto and RaftProto can share that code rather than duplicate it
* Publish RaftProto and KvProto on crates.io

How does that sound? @BusyJay @Hoverbear

**Comment by BusyJay** [comment]
It sounds good to me. But I have also been thinking about removing raft dependency from kvproto, so that raft is free to change its protocols, or even no protocols at all.

/cc @siddontang what do you think?

**Comment by ice1000** [comment]
> so that raft is free to change its protocols, or even no protocols at all.

I suppose raft and kvproto will depend on the same "raftproto" library mentioned in @nrc's comment

**Comment by Hoverbear** [comment]
@nrc I approve of this idea!

**Comment by nrc** [comment]
> It sounds good to me. But I have also been thinking about removing raft dependency from kvproto, so that raft is free to change its protocols, or even no protocols at all.

@BusyJay @siddontang do you have any more thoughts on this?

**Comment by BusyJay** [comment]
I just take a look into the protocols, it seems there are many places using eraft protos. It's not practical to remove the dependency now. So your proposal seems the only way to publish.

**Comment by nrc** [comment]
Cool, thanks for looking at the protocols!

**Comment by Daniel-B-Smith** [comment]
Any update on publishing the crate?

**Comment by Hoverbear** [comment]
@Daniel-B-Smith We're currently not able to publish this due to some ongoing work by @nrc . We're planning to publish this crate (along with our Rust client) in Q4. Sorry for the delays.

https://github.com/pingcap/kvproto/blob/121d118fa10e253013b7dc03d0d5b9148f859b78/Cargo.toml#L31-L32

**Comment by Daniel-B-Smith** [comment]
Thanks for the update!

**Comment by xxchan** [comment]
Any update?

**Comment by disksing** [comment]
any one know the status of this issue?  cc @sticnarf @BusyJay

**Comment by BusyJay** [comment]
kvproto now uses forked rust-protobuf, I believe forked rust-protobuf needs to be published first.

**Comment by BusyJay** [comment]
As far as I know, no one is working on it.

**Comment by BusyJay** [comment]
As far as I know, no one is working on it.

**Comment by xxchan** [comment]
Maybe related post:  [Move kvproto to tikv org](https://internals.tidb.io/t/topic/67)
